src/apprt/gtk-ng/class/command_palette.zig

```diff
@@ -0,0 +1,426 @@
+    /// returned will be floating.
+    pub fn new(window: *Window, config: *Config) ?*Self {
+        if (comptime !Self.isSupported()) return null;
+
+        const self = gobject.ext.newInstance(Self, .{});
+
+        const app = Application.default();
+        const priv = self.private();
+
+        // The config that we will use to begin with. It should already have
+        // had it's reference count increased, and we take ownership of it.
+        priv.config = config;
+
+        // Note that we _do not_ take a reference to the window as this can
+        // cause a circular reference count, which means that the window/command
+        // palette pair will never be destroyed`.
+        priv.window = window;
+
+        // Initialize an arena to make cleaning up our allocations easier.
+        priv.arena = .init(app.core().alloc);
+
+        // Ensure that we always have the same config as the window.
+        _ = gobject.Object.bindProperty(
+            window.as(gobject.Object),
+            "config",
+            self.as(gobject.Object),
+            "config",
+            .{},
+        );
+
+        self.updateConfig();
+
+        return self;
+    }
+
+    //---------------------------------------------------------------
+    // Virtual Methods
+
+    fn init(self: *Self, _: *Class) callconv(.c) void {
+        gtk.Widget.initTemplate(self.as(gtk.Widget));
+    }
+
+    fn dispose(self: *Self) callconv(.c) void {
+        const priv = self.private();
+
+        priv.source.removeAll();
+        priv.arena.deinit();
+
+        if (priv.config) |config| {
+            config.unref();
+            priv.config = null;
+        }
+
+        gtk.Widget.disposeTemplate(
+            self.as(gtk.Widget),
+            getGObjectType(),
+        );
+
+        gobject.Object.virtual_methods.dispose.call(
+            Class.parent,
+            self.as(Parent),
+        );
+    }
+
+    //---------------------------------------------------------------
+    // Signal Handlers
+
+    fn propConfig(self: *CommandPalette, _: *gobject.ParamSpec, _: ?*anyopaque) callconv(.c) void {
+        self.updateConfig();
+    }
+
+    fn searchStopped(_: *gtk.SearchEntry, self: *CommandPalette) callconv(.c) void {
+        // ESC was pressed - close the palette
+        const priv = self.private();
+        _ = priv.dialog.close();
+    }
+
+    fn searchActivated(_: *gtk.SearchEntry, self: *CommandPalette) callconv(.c) void {
+        // If Enter is pressed, activate the selected entry
+        const priv = self.private();
+        self.activated(priv.model.getSelected());
+    }
+
+    fn rowActivated(_: *gtk.ListView, pos: c_uint, self: *CommandPalette) callconv(.c) void {
+        self.activated(pos);
+    }
+
+    //---------------------------------------------------------------
+
+    pub fn toggle(self: *CommandPalette) void {
+        const priv = self.private();
+        const window = priv.window orelse {
```

Comment:
Hm, just thinking out loud here, I wonder if we should associate the command palette with a _surface_ instead. That's how it works on macOS and this lets the command palette remain opening when switching tabs. I'm not sure if that'd translate well to GTK so this is more of an open thought.

Reply:
That would mean a lot more command palette objects laying around. Right now we have one per window and it lasts the lifetime of the window. Doing it per surface means a lot more long-lived objects. Plus the dialog is modal to the window so you can't really be switching tabs/splits while the palette is open.

Reply:
On macOS we dispose of the entire object whenever it's closed so you only have as many active as are open. I think we should do this here too (regardless of our target). 

> Plus the dialog is modal to the window so you can't really be switching tabs/splits while the palette is open.

Ah, yeah this would be a limitation. I wonder if longer term we should change it to an overlay item.

Reply:
I've changed the command palette so that it's dynamically allocated as needed, and then disposed of. Attaching it to a surface would be possible in the future, but right now there is no code for getting the window that a surface is attached to (at least not that I can find).

Reply:
FWIW adw.Dialog can only ever be attached to a window (the present method technically takes any widget but the dialog will just attach itself to the window that contains said widget), so it seems like a fundamental limitation unless we switch to use overlays. But then, I'm not sure if that's a good idea since we do have a lot of information to display and restricting ourselves to only the area of the split might only complicate matters